iron/gpg-packet.c

```diff
@@ -0,0 +1,1150 @@
+ */
+void
+generate_gpg_public_key_packet(const Key * ssh_key, gpg_packet * pkt)
+{
+    pkt->tag = GPG_TAG_PUBLIC_KEY;
+    pkt->data = sshbuf_new();
+
+    sshbuf_put_u8(pkt->data, GPG_KEY_VERSION);
+    sshbuf_put_u32(pkt->data, iron_gpg_now());
+    sshbuf_put_u8(pkt->data, GPG_PKALGO_RSA_ES);
+    iron_put_bignum(pkt->data, ssh_key->rsa->n);
+    iron_put_bignum(pkt->data, ssh_key->rsa->e);
+    pkt->len = sshbuf_len(pkt->data);
+}
+
+/**
+ *  Generate GPG Subkey Packet for cv25519 key.
+ *
+ *  Format a GPG Subkey packet containing a cv25519 public key.
+ *
+ *  @param pub_key Byte array containing cv25519 public key
+ *  @param pk_len Num bytes in pub_key
+ *  @param pkt Place to put generated packet. Caller should sshbuf_free pkt->data
+ */
+void
+generate_gpg_curve25519_subkey_packet(const u_char * pub_key, size_t pk_len, gpg_packet * pkt)
+{
+    pkt->tag = GPG_TAG_PUBLIC_SUBKEY;
+    pkt->data = sshbuf_new();
+
+    sshbuf_put_u8(pkt->data, GPG_KEY_VERSION);
+    sshbuf_put_u32(pkt->data, iron_gpg_now());
+    sshbuf_put_u8(pkt->data, GPG_PKALGO_ECDH);      //  Curve25519 is an instance of ECDH
+    generate_gpg_curve25519_pubkey_parms(pub_key, pk_len, pkt->data);
+    pkt->len = sshbuf_len(pkt->data);
+}
+
+/**
+ *  Generate GPG User ID packet.
+ *
+ *  @param user_id String identifying user (name and <email>, often)
+ *  @param pkt Place to write packet. Caller should sshbuf_free pkt->data
+ */
+void
+generate_gpg_user_id_packet(const char * user_id, gpg_packet * pkt)
+{
+    pkt->tag = GPG_TAG_USERID;
+    pkt->data = sshbuf_from(user_id, strlen(user_id));
+    pkt->len = sshbuf_len(pkt->data);
+}
+
+/**
+ *  Write the fixed info at the start of the signature packet body.
+ *
+ *  @param body Place to write body data
+ *  @param sig_class Type of signature to generate
+ *  @param pubkey_tag Tag from the public key or subkey packet preceding signature
+ */
+static void
+populate_signature_header(struct sshbuf * body, int sig_class, gpg_tag pubkey_tag)
+{
+    sshbuf_put_u8(body, GPG_SIG_VERSION);
+    sshbuf_put_u8(body, sig_class);
+    sshbuf_put_u8(body, GPG_PKALGO_RSA_ES);
+    sshbuf_put_u8(body, GPG_HASHALGO_SHA256);
+    sshbuf_put_u16(body, 24);       //  Length of hashed subpackets
+    sshbuf_put_u8(body, 5);         //  Length of signature creation time subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_SIGNATURE_CREATION_TIME);
+    sshbuf_put_u32(body, iron_gpg_now());
+    sshbuf_put_u8(body, 5);         //  Length of key lifetime subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_KEY_LIFETIME);
+    sshbuf_put_u32(body, 0);            //  Does not expire
+    sshbuf_put_u8(body, 5);         //  Length of preferred symmetric algorithm subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_PREF_SYM_ALGO);
+    sshbuf_put_u8(body, GPG_SKALGO_AES256);
+    sshbuf_put_u8(body, GPG_SKALGO_AES192);
+    sshbuf_put_u8(body, GPG_SKALGO_AES128);
+    sshbuf_put_u8(body, 0);         //  No fourth preferred SK algorithm
```

Comment:
Yes - it's a fixed array of up to four preferred algorithms. I think there are similar arrays for PK algorithms and hash algorithms, and maybe for compression algorithms.